packages/flutter/lib/src/rendering/object.dart

```diff
@@ -1111,6 +1111,19 @@ class PipelineOwner {
 /// The [RenderBox] subclass introduces the opinion that the layout
 /// system uses Cartesian coordinates.
 ///
+/// ## Lifecycle
+///
+/// A [RenderObject] must [dispose] when it is no longer needed. The creator
+/// of the object is responsible for disposing of it. Typically, the creator is
+/// a [RenderObjectElement], and that element will dispose the object it creates
+/// when it is unmounted.
+///
+/// [RenderObject]s are responsible for cleaning up any expensive resources
+/// they hold when [dispose] is called, such as [Picture] or [Image] objects.
+/// They must _only_ dispose of their [layer] object if [isRepaintBoundary] is
```

Comment:
This is not super-clear to me. If I own the layer (e.g. because I set it in my paint method) why am I not in charge of releasing it as well? How can my parent still use it if I (the owner) are no longer arround?

Reply:
@dnfield and I haven't finished this discussion, so I'm also interested in the ownership model here. I think there's shared ownership going on, but we lack means to express it. In a ref-counted language I'd be thinking about it as follows: a layer's ref count is bumped by its parent layer and its render object. The count can be in the following states:

2 - when a layer has both a parent layer and an owning RenderObject
1 - when a layer has either a parent layer or an owning RenderObject, but not both
0 - when nothing references the layer, this is when I'd clear the resources of this layer, this is also when recursively this layer's descendants' ref-counts are dropped by one and, possibly, reach 0, get their resources cleaned up, and so on.

This way we don't need to treat repaint boundaries as special. Any render object can own a sub-tree of layers, and the recursive disposal of render objects introduced by this PR makes sure we don't leak layers.

Reply:
Yes, this doc is incorrect and I have updated it.

I may have to revisit this as I continue to work out the details about layer disposal, but I think it makes more sense now. PTAL.

Reply:
Thanks. The doc comment is clear, but I'm not sure if what it describes is the right thing to do. Perhaps it is, and I just don't have a full picture in my head. Here's a couple of examples that I'm not sure are handled properly:

- A repaint boundary has a `RenderOpacity` child. The repaint boundary disposes of its layer. Is the `RepaintBoundary` pointing to a disposed layer (i.e. a dangling pointer)?
- A repaint boundary has another repaint boundary as a child. The parent disposes of its layer. Do we recursively delete pictures of the child repaint boundary? If not, how do we know to not delete them? If yes, how does the child repaint boundary know what its resources got cleaned-up and it needs to repaint?

Reply:
I've changed this wording because it's no longer relevant to this patch. This patch doesn't add anything for disposing of layers, that will be coming in a separate patch.

To start to answer these questions though - if a repaint boundary disposes its layer, it means that repaint boundary is going way and won't be drawn again without getting completely recreated/repainted. And so if it has any child repaint boundaries that have not been moved to some active part of the tree, they will not get drawn again either unless recreated/repainted.